### PR TITLE
Extended Gemm interface to support mixed precision operations

### PR DESCRIPTION
**Pull request by OuadiElfarouki**
Extend Gemm operator interface to support mixed precision operations, namely by decoupling matrix A and B type `element_in_t` from output matrix C and scalars alpha and beta type `element_out_t`. 

Following oneMKL's spec notation for Gemm API : https://spec.oneapi.io/versions/latest/elements/oneMKL/source/domains/blas/gemm.html#onemkl-blas-gemm, this PR enables (Ta==Tb) to be set independently from (Tc==Ts). This feature has been enabled at a first stage for Ta=Tb=sycl::half and Tc=Ts=float. Thus enabling half support also enables the mixed precision case of (half, float) for gemm. 

Changes include:
- Updating different Gemm kernel implementations to account for the decoupled types.
- Necessary CMake and Kernel generation scripts updates to account for the **couple of types** instead of **single type** in gemm case.
- Necessary changes to unit-tests to account for this feature in the Gemm case.

Note : 
Following oneMKL expected Gemm API, Support of `bfloat16-float ` and `std::int8_t-float` would be straightforward afterwards, but the additional cases of `Ta==Tb==Tc` while `Ts` _(alpha & beta)_ is separate will require additional decoupling & re-design work..

### CODE REVIEW

File: src/operations/blas3/gemm_local_joint_matrix.hpp [hunk 1]
```diff
@@ -62,6 +62,7 @@ namespace blas {
  * @tparam TransA  iff true, matrix A will be transposed on the fly
  * @tparam TransB  iff true, matrix B will be transposed on the fly
  * @tparam element_t  type of matrix elements
+ * @tparam element_out_t  type of output matrix elements (C) and scalars
```
Comment (hjabird) [inline src/operations/blas3/gemm_local_joint_matrix.hpp:65]: I'd also like to see `element_t` renamed to `element_in_t`.

File: src/sb_handle/portblas_handle.hpp [hunk 2]
```diff
@@ -336,7 +339,7 @@ inline typename SB_Handle::event_t SB_Handle::execute(
   /* Create the cube buffer that will hold the output of the partial gemm */
   auto cube_buffer = acquire_temp_mem < is_usm ? helper::AllocType::usm
                                                : helper::AllocType::buffer,
-       element_t > (rows * cols * depth);
+       element_out_t > (rows * cols * depth);
```
Comment (hjabird) [inline src/sb_handle/portblas_handle.hpp:342]: I know this code that's being updated, but it looks like a type is being compared to a value!? What's going on?

File: test/blas_test_macros.hpp [hunk 3]
```diff
@@ -157,7 +204,8 @@
                                         name_generator);
 #endif  // BLAS_ENABLE_COMPLEX
 
-/** Registers test for all supported data types
+/** Registers test for all supported data
+ * types
```
Comment (hjabird) [inline test/blas_test_macros.hpp:208]: Why this change?

File: test/unittest/blas3/blas3_gemm_common.hpp [hunk 4]
```diff
@@ -86,7 +86,14 @@ inline std::vector<scalar_t> interleaved_to_strided(
   return output;
 }
 
-template <typename scalar_t, helper::AllocType mem_alloc>
+/**
+ * @brief verify gemm correctness against reference BLAS.
+ *
+ * @tparam scalar_in_t  type of input matrices elements (A, B)
+ * @tparam scalar_t  type of output matrix elements (C) and scalars
+ * (gemm_arguments)
```
Comment (hjabird) [inline test/unittest/blas3/blas3_gemm_common.hpp:94]: `scalar_out_t`?

File: src/operations/blas3/gemm_interleaved.hpp [hunk 5]
```diff
@@ -486,28 +493,41 @@ class Gemm<input_t, output_t, /* DoubleBuffer = */ false, /* NbcA = */ false,
...
       for (int j = 0; j < item_rows; ++j) {
 #pragma unroll
         for (int b = 0; b < item_batchs / VectorSize; ++b) {
 #ifdef __ADAPTIVECPP__
-          if constexpr (is_half<element_t>::value) {
-            *reg_res = reg_a[j * (item_batchs / VectorSize) + b] *
-                           reg_b[i * (item_batchs / VectorSize) + b] +
-                       *reg_res;
+          if constexpr (is_half<element_in_t>::value ||
+                        !std::is_same_v<element_in_t, element_out_t>) {
```
Comment (s-Nick) [inline src/operations/blas3/gemm_interleaved.hpp:505]: you check opposite condition, why not checking the same condition and keeping the order of if/else branch code. I think it would be more readable.

File: src/operations/blas3/gemm_interleaved.hpp [hunk 6]
```diff
@@ -486,28 +493,41 @@ class Gemm<input_t, output_t, /* DoubleBuffer = */ false, /* NbcA = */ false,
...
+#pragma unroll
+            for (int v = 0; v < VectorSize; ++v) {
```
Comment (s-Nick) [inline src/operations/blas3/gemm_interleaved.hpp:507]: Can you elaborate why this loop is necessary?

**Review by s-Nick — COMMENTED** [review]
Thank you for you work @OuadiElfarouki.
There is some confusion with `element_in_t` and `element_out_t` because you forget to update `element_t` to one or the other. Can you please fix it? @hjabird highlighted many of them. In general, I agree that something like `in_element_t` and `out_element_t` would be more readable, please consider it.

File: src/sb_handle/portblas_handle.hpp (hunk 2 above)
Comment (hjabird) [reply src/sb_handle/portblas_handle.hpp:342]: @s-Nick has explained this to me.

File: cmake/CmakeFunctionHelper.cmake [hunk 7]
```diff
@@ -702,7 +722,17 @@ else() # default cpu backend
...
-  endforeach()
+  
+    if(BLAS_ENABLE_HALF)
+      add_gemm_configuration(
+        "half"  128 "false" "false" "false"
+        64 4 4 8 8 1 1 1 1 1 1 1 1 1 float float "no_local" "standard" "full" 1 "strided" "false" "false")
+      add_gemm_configuration(
+        "half" 64 "false" "false" "false"
+        64 2 2 4 4 1 1 1 1 4 4 1 1 1 float float "no_local" "standard" "full" 4 "interleaved" "false" "false")
+    endif()
```
[the remark this answers is not in the document]
Comment (hjabird) [reply cmake/CmakeFunctionHelper.cmake:733]: @s-Nick has explained this to me via other channels.

File: src/operations/blas3/gemm_interleaved.hpp (hunk 5 above)
Comment (OuadiElfarouki) [reply src/operations/blas3/gemm_interleaved.hpp:505]: Agree @s-Nick. I avoided changing the scalar_t and element_t before to make less changes xD But I agree with you guys @hjabird it would improve readability to have same "format". So will push the renaming changes next.

File: src/operations/blas3/gemm_interleaved.hpp (hunk 6 above)
Comment (OuadiElfarouki) [reply src/operations/blas3/gemm_interleaved.hpp:507]: @s-Nick AdaptiveCpp doesn't support sycl::vec<sycl::half> for mad operation, so had to break it down to its elements through the loop

File: include/operations/blas3_trees.h [hunk 8]
```diff
@@ -169,7 +169,8 @@ struct Tile {
  * @tparam TransB  iff true, matrix B will be transposed on the fly
  * @tparam SymmA   whether the matrix A is a symmetric triangular matrix
  * @tparam SymmB   whether the matrix B is a symmetric triangular matrix
- * @tparam element_t  type of matrix elements
+ * @tparam element_in_t  type of input matrix elements (A & B)
+ * @tparam element_t  type of output matrix elements (C) and scaling parameters
```
[the remark this answers is not in the document]
Comment (OuadiElfarouki) [reply include/operations/blas3_trees.h]: Thanks for the hint @pgorlani, changes have been reduced now!

File: src/operations/blas3/gemm_common.hpp [hunk 9]
```diff
@@ -42,13 +42,22 @@ mul_add(T a, T b, T c,
...
+                                is_sycl_scalar<Tin>::value>::type * = 0) {
   return (cl::sycl::mad(a, b, c));
 }
 
+template <typename Tin, typename Tout>
+static PORTBLAS_INLINE Tout
+mul_add(Tin a, Tin b, Tout c,
+        typename std::enable_if<!std::is_same<Tin, Tout>::value &&
+                                is_sycl_scalar<Tin>::value>::type * = 0) {
+  return static_cast<Tout>(a) * static_cast<Tout>(b) + c;
```
[the remark this answers is not in the document]
Comment (OuadiElfarouki) [reply src/operations/blas3/gemm_common.hpp:58]: @hjabird yes indeed, it should be actually done in `Tin` but I had to fall back to `Tout` because gemm unit-tests failed on the oneMKL side (portBLAS backend) due to precision mismatches. This needs to be investigated though I agree, as casting to float beats the purpose of having half mul.

File: src/operations/blas3/gemm_local_joint_matrix.hpp (hunk 1 above)
Comment (OuadiElfarouki) [reply src/operations/blas3/gemm_local_joint_matrix.hpp]: The extra template parameter has been removed.

File: include/operations/blas3_trees.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (pgorlani) [reply include/operations/blas3_trees.h]: Thank you so much, @OuadiElfarouki!

File: test/blas_test_macros.hpp (hunk 3 above)
Comment (OuadiElfarouki) [reply test/blas_test_macros.hpp:208]: Formatting mistake, will revert it!

File: src/operations/blas3/gemm_common.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (hjabird) [reply src/operations/blas3/gemm_common.hpp:58]: Do we have any ticket or issue to track that this needs to be investigated?

File: test/blas_test_macros.hpp (hunk 3 above)
Comment (hjabird) [reply test/blas_test_macros.hpp]: Fixed in [c7aedc9](https://github.com/codeplaysoftware/portBLAS/pull/500/commits/c7aedc965dae663ae2ee9c171f69302c38a4d092).

File: test/unittest/blas3/blas3_gemm_common.hpp (hunk 4 above)
Comment (hjabird) [reply test/unittest/blas3/blas3_gemm_common.hpp]: Appears to be done.

File: src/operations/blas3/gemm_common.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (OuadiElfarouki) [reply src/operations/blas3/gemm_common.hpp:58]: @hjabird not that I know of no, I've put a placeholder [oneMKL PR](https://github.com/OuadiElfarouki/oneMKL/pull/1) locally that we can open once the portBLAS PRs are merged. Something similar was raised in this [PR](https://github.com/oneapi-src/oneMKL/pull/466/files#r1552523436) regarding the test precision tolerance. So basically if the precision on oneMKL tests is handled loosely for half/mixed operations, we can drop this static_cast on portBLAS side as it only causes failures on oneMKL tests not on our portBLAS unittests.

File: test/unittest/blas3/blas3_gemm_common.hpp (hunk 4 above)
Comment (OuadiElfarouki) [reply test/unittest/blas3/blas3_gemm_common.hpp]: @hjabird yes done!

**Comment by OuadiElfarouki** [comment]
@s-Nick @Rbiessy @hjabird I've rebased and fixed some bugs in the current PR after recent commits. Feel free to check it ! Thanks.

**Review by hjabird — APPROVED** [review]
